Add unit to RuntimeMetric

zacw7 opened the pull request:
Resolves [#17633](https://github.com/prestodb/presto/issues/17633)
Changing the runtimeStats structure from a map to an array, which is incompatible with older version of Presto CLI/Verifier.
```JSON
"runtimeStats": [
  {
    "name": "getTableHandleTimeNanos",
    "unit": "NANO",
    "sum": 42914,
    "count": 1,
    "max": 42914,
    "min": 42914
  },
  {
    "name": "S1-storageReadDataBytes",
    "unit": "BYTE",
    "...": "..."
  },
  "...": "..."
]
```

Resubmit the PR as the inconsistency has been fixed on the CPP side.
```
== RELEASE NOTES ==

General Changes
* Add the support for setting units for runtime metrics. The older version of Presto CLI and Verifier will be incompatible.
```

aweisberg commented:
Is this going to make verifier not work still? I am asking because we rely on it for release verification.

zacw7 commented:
> Is this going to make verifier not work still? I am asking because we rely on it for release verification.

Yes :( But we are not planning to land this in the current release since it's too close to the release cut. Will try to figure it out how to handle the verifier failure next week.

kewang1024 commented on presto-common/src/main/java/com/facebook/presto/common/RuntimeMetric.java:
```diff
@@ -38,29 +39,49 @@
      * Creates a new empty RuntimeMetric.
      *
-     * @param name Name of this metric. If used in the presto core code base, this should be a value defined in {@link RuntimeMetricName}. But connectors could use arbitrary names.
+     * @param name Name of this metric. If used in the presto core code base, this should be a value defined in {@link RuntimeMetricKey}. But connectors could use arbitrary names.
```
Also add @param for unit

kewang1024 commented on presto-common/src/main/java/com/facebook/presto/common/RuntimeMetricKey.java:
```diff
@@ -0,0 +1,111 @@
+@ThriftStruct
+public class RuntimeMetricKey
+{
+    public static final RuntimeMetricKey DRIVER_COUNT_PER_TASK = new RuntimeMetricKey("driverCountPerTask");
+    public static final RuntimeMetricKey TASK_ELAPSED_TIME_NANOS = new RuntimeMetricKey("taskElapsedTimeNanos");
+    public static final RuntimeMetricKey OPTIMIZED_WITH_MATERIALIZED_VIEW_COUNT = new RuntimeMetricKey("optimizedWithMaterializedViewCount");
+    public static final RuntimeMetricKey MANY_PARTITIONS_MISSING_IN_MATERIALIZED_VIEW_COUNT = new RuntimeMetricKey("manyPartitionsMissingInMaterializedViewCount");
+    public static final RuntimeMetricKey SKIP_READING_FROM_MATERIALIZED_VIEW_COUNT = new RuntimeMetricKey("skipReadingFromMaterializedViewCount");
+    public static final RuntimeMetricKey FRAGMENT_RESULT_CACHE_HIT = new RuntimeMetricKey("fragmentResultCacheHitCount");
+    public static final RuntimeMetricKey FRAGMENT_RESULT_CACHE_MISS = new RuntimeMetricKey("fragmentResultCacheMissCount");
+    public static final RuntimeMetricKey GET_VIEW_TIME_NANOS = new RuntimeMetricKey("getViewTimeNanos");
+    public static final RuntimeMetricKey GET_MATERIALIZED_VIEW_TIME_NANOS = new RuntimeMetricKey("getMaterializedViewTimeNanos");
+    public static final RuntimeMetricKey GET_MATERIALIZED_VIEW_STATUS_TIME_NANOS = new RuntimeMetricKey("getMaterializedViewStatusTimeNanos");
+    public static final RuntimeMetricKey GET_TABLE_HANDLE_TIME_NANOS = new RuntimeMetricKey("getTableHandleTimeNanos");
+    public static final RuntimeMetricKey GET_TABLE_METADATA_TIME_NANOS = new RuntimeMetricKey("getTableMetadataTimeNanos");
+    public static final RuntimeMetricKey GET_SPLITS_TIME_NANOS = new RuntimeMetricKey("getSplitsTimeNanos");
+    public static final RuntimeMetricKey LOGICAL_PLANNER_TIME_NANOS = new RuntimeMetricKey("logicalPlannerTimeNanos");
+    public static final RuntimeMetricKey FRAGMENT_PLAN_TIME_NANOS = new RuntimeMetricKey("fragmentPlanTimeNanos");
+    public static final RuntimeMetricKey GET_LAYOUT_TIME_NANOS = new RuntimeMetricKey("getLayoutTimeNanos");
+    public static final RuntimeMetricKey REWRITE_AGGREGATION_IF_TO_FILTER_APPLIED = new RuntimeMetricKey("rewriteAggregationIfToFilterApplied");
+    // Time between task creation and start.
+    public static final RuntimeMetricKey TASK_QUEUED_TIME_NANOS = new RuntimeMetricKey("taskQueuedTimeNanos");
+    // Total operation time of a task on a worker. TASK_ELAPSED_TIME_NANOS - TASK_SCHEDULED_TIME_NANOS is the time when the task is doing nothing, e.g. it might be waiting for splits/inputs.
+    public static final RuntimeMetricKey TASK_SCHEDULED_TIME_NANOS = new RuntimeMetricKey("taskScheduledTimeNanos");
+    // Blocked time for the operators due to waiting for inputs.
+    public static final RuntimeMetricKey TASK_BLOCKED_TIME_NANOS = new RuntimeMetricKey("taskBlockedTimeNanos");
+    // Time taken for a read call to storage
+    public static final RuntimeMetricKey STORAGE_READ_TIME_NANOS = new RuntimeMetricKey("storageReadTimeNanos");
+    // Size of the data retrieved by read call to storage
+    public static final RuntimeMetricKey STORAGE_READ_DATA_BYTES = new RuntimeMetricKey("storageReadDataBytes");
+
```
Why not create them with unit as well?

kewang1024 commented on presto-common/src/main/java/com/facebook/presto/common/RuntimeMetricKey.java:
```diff
@@ -0,0 +1,111 @@
+
+    @JsonCreator
+    @ThriftConstructor
+    public RuntimeMetricKey(@JsonProperty("name") String name, @JsonProperty("unit") RuntimeUnit unit)
```
Why do we add unit to both `RuntimeMetricKey` and `RuntimeMetric`?

kewang1024 reviewed (commented):
Can we explain the motivation behind the change from map to list and what's the benefit ?

kewang1024 commented on presto-common/src/main/java/com/facebook/presto/common/RuntimeUnit.java:
```diff
@@ -0,0 +1,36 @@
+@ThriftEnum
+public enum RuntimeUnit
+{
+    NONE(0), NANO(1), BYTE(2);
```
1. Can we have one more `TIME` to represent how many times 
2. Can we remove `NONE`, why do we need it?

zacw7 replied:
> Can we have one more TIME to represent how many times

You mean use `TIME` for the counting metrics like `driverCountPerTask ` or `fragmentResultCacheHitCount`? I think it might be a little be confusing. `TIME` looks like a time unit. Also the current setup is align with [Velox](https://github.com/facebookincubator/velox/blob/fdbbfd2354075da80e98f18055d744172f15e15f/velox/common/base/RuntimeMetrics.h#L25). I think it's good for start.

> Can we remove NONE, why do we need it?
 
I would suggest keeping it:
1. For compatibility reasons, we are introducing this unit to runtime metrics and there are many metrics haven't been set up with units yet, we would like those metrics to have a default fall-back value `NONE`.
2. For some metrics like `optimizedWithMaterializedView` ( _1 means the query is rewritten to read from available materialized views to improve the performance_), it does make much sense to set a unit to them, so let's keep it as `NONE`.

kewang1024 replied:
Sounds good, can we 
1. Link the Velox file in the description as well since those two have to be compatible with each other
2. Can we add an issue or task as follow up to slowly deprecate the NONE and adopt meaningful units
3. As for `TIME`, I understand the confusion (I would add a comment to avoid confusion), but also hard to think of any other names :) , feel free to change to a name more appropriate,  let's include this unit as an option in the follow up task/issue as well

kewang1024 commented on presto-common/src/main/java/com/facebook/presto/common/RuntimeStats.java:
```diff
@@ -87,23 +89,33 @@ public Map<String, RuntimeMetric> getMetrics()
 
     public void addMetricValue(String name, long value)
```
Can we remove this method and migrate all the calls to `addMetricValue `?
Keep this method just makes the system very prone to errors

kewang1024 commented on presto-common/src/main/java/com/facebook/presto/common/RuntimeStats.java:
```diff
@@ -87,23 +89,33 @@ public Map<String, RuntimeMetric> getMetrics()
     }
 
     public void addMetricValueIgnoreZero(String name, long value)
+    {
```
Same

kewang1024 commented on presto-common/src/main/java/com/facebook/presto/common/RuntimeMetric.java:
```diff
@@ -30,6 +31,7 @@
 public class RuntimeMetric
 {
     private final String name;
```
Do we still need the `name` here since it's duplicate information?

zacw7 replied:
I'm kinda cautious to remove the existing field from the protocol. Some components might rely on it.

kewang1024 replied:
Looks like the [Velox](https://github.com/facebookincubator/velox/blob/fdbbfd2354075da80e98f18055d744172f15e15f/velox/common/base/RuntimeMetrics.h#L25) definition doesn't have it, so to make it more consistent, can we at least check what components are relying on it and would it be easy to remove?

zacw7 replied:
I'm afraid this is another breaking change. The main reason we choose to keep the map structure instead of converting into array is that we cannot remove the existing information.

kewang1024 commented on presto-common/src/main/java/com/facebook/presto/common/RuntimeMetric.java:
```diff
@@ -39,28 +41,46 @@
             @JsonProperty("min") long min)
     {
-        this(name);
+        this(name, unit);
```
Since the older version would pass in null as the `unit`, would it throw exceptions because of the `requireNonNull ` check in construction function?

zacw7 replied:
It will be checked inside `this(name, unit)`.

kewang1024 replied:
Yes, and it throws exception, so wouldn't the new version of presto that has this change be incompatible with older version?

zacw7 commented:
Hey @kewang1024, thanks for reviewing the PR. I've addressed most of the comments and removed the error-prone methods like `addMetricValue(String name, long value)` so `unit` must be specified. For now, most of the metrics are just set with `NONE`. There is another work planned for Eric to assign the right units to the metrics and to make better use of the units.

kewang1024 commented:
> Hey @kewang1024, thanks for reviewing the PR. I've addressed most of the comments and removed the error-prone methods like `addMetricValue(String name, long value)` so `unit` must be specified. For now, most of the metrics are just set with `NONE`. There is another work planned for Eric to assign the right units to the metrics and to make better use of the units.

Thanks! For the work planned, can we create another task or issue on github for better tracking?

zacw7 commented:
> > Hey @kewang1024, thanks for reviewing the PR. I've addressed most of the comments and removed the error-prone methods like `addMetricValue(String name, long value)` so `unit` must be specified. For now, most of the metrics are just set with `NONE`. There is another work planned for Eric to assign the right units to the metrics and to make better use of the units.
> 
> Thanks! For the work planned, can we create another task or issue on github for better tracking?

https://github.com/prestodb/presto/issues/17770

kewang1024 reviewed (approved):
I took a second look
1. The reason why although it adds a new field but it's not breaking when interacting with old version presto client is that the data is sent one-way from coordinator to client (thus adding `RuntimeUnit` and having its non-null check is fine)

2. Why we can't remove redundant `name` from `RuntimeMetric` right now is presto client is reading this information
`StatusPrinter`:
`stats.getRuntimeStats().getMetrics().values().stream().sorted(Comparator.comparing(RuntimeMetric::getName))`

    However, this would cause a potential bug when integrating with Velox since [Velox](https://github.com/facebookincubator/velox/blob/fdbbfd2354075da80e98f18055d744172f15e15f/velox/common/base/RuntimeMetrics.h#L25) doesn't have `name` in `RuntimeMetric`
So when data is collected from c++ worker to java coordinator, `name` would be missing from `RuntimeMetric`

    Suggestion: make changes to `StatusPrinter` to use `name` from the map instead of from `RuntimeMetric`

    I'm fine with either doing it in this PR or in the future PR (create an issue for it)
    cc: @yuanzhanhku

yuanzhanhku commented on presto-common/src/main/java/com/facebook/presto/common/RuntimeMetric.java:
```diff
@@ -85,6 +90,13 @@ public String getName()
     }
 
+    @JsonProperty
+    @ThriftField(2)
```
Is there a specific reason to reuse an existing thrift id? 
When adding a new field in Thrift, we should use a new id. Otherwise, existing serialized Thrift struct cannot be parsed. This will also cause compatibility issue if the coordinator and the worker run different versions.

yuanzhanhku commented on presto-hive-common/src/main/java/com/facebook/presto/hive/HiveFileContext.java:
```diff
@@ -82,7 +83,7 @@ public Optional<Long> getFileSize()
     public void incrementCounter(String name, long value)
```
We may need to add `unit` to this function and migrate existing usages.
This can be done in a separate PR.

zacw7 replied:
Sounds good. Created an issue to keep track of it - https://github.com/prestodb/presto/issues/17777

yuanzhanhku commented on presto-main/src/main/java/com/facebook/presto/execution/StageExecutionInfo.java:
```diff
@@ -175,11 +180,11 @@ public static StageExecutionInfo create(
-            mergedRuntimeStats.addMetricValueIgnoreZero(RuntimeMetricName.TASK_BLOCKED_TIME_NANOS, taskStats.getTotalBlockedTimeInNanos());
+            mergedRuntimeStats.addMetricValue(DRIVER_COUNT_PER_TASK, NONE, taskStats.getTotalDrivers());
+            mergedRuntimeStats.addMetricValue(TASK_ELAPSED_TIME_NANOS, NONE, taskStats.getElapsedTimeInNanos());
+            mergedRuntimeStats.addMetricValueIgnoreZero(TASK_QUEUED_TIME_NANOS, NONE, taskStats.getQueuedTimeInNanos());
```
Are you planning to change the units in a separate PR?

zacw7 replied:
Yes

rschlussel commented:
@zacw7 this is causing failures for verifier.  What's the plan here?

zacw7 commented:
> @zacw7 this is causing failures for verifier. What's the plan here?

It seems it only affects Sapphire verifier, let me test if relaxing the null check can prevent the error.

zacw7 commented:
> @zacw7 this is causing failures for verifier. What's the plan here?

Fixed in https://github.com/prestodb/presto/pull/17840
